controllers/oauth_client.go

```diff
@@ -0,0 +1,85 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"html/template"
+	"net/http"
+
+	service "go-crawler-challenge/services/oauth"
+
+	"github.com/beego/beego/v2/server/web"
+)
+
+// OauthClientController operations for oauth client generator
+type OauthClientController struct {
+	BaseController
+}
+
+// NestPrepare prepares some configurations to the controller
+func (c *OauthClientController) NestPrepare() {
+	c.actionPolicyMapping()
+}
+
+// URLMapping maps oauth client controller actions to functions
+func (c *OauthClientController) URLMapping() {
+	c.Mapping("New", c.New)
+	c.Mapping("Create", c.Create)
+}
+
+// actionPolicyMapping maps oauth client controller actions to policies
+func (c *OauthClientController) actionPolicyMapping() {
+	c.MappingPolicy("New", Policy{requireAuthenticatedUser: true})
+	c.MappingPolicy("Create", Policy{requireAuthenticatedUser: true})
+}
+
+// New handles a form for generating a new oauth client credential and only display into the same form
+// @Title New
+// @Description show an oauth client credential form with generated values
+// @Success 200
+// @router /oauth_client [get]
+func (c *OauthClientController) New() {
+	web.ReadFromRequest(&c.Controller)
+	flash := web.NewFlash()
+
+	clientId := c.GetString("client_id")
+	if clientId != "" {
+		oauthClient, err := service.ClientStore.GetByID(context.TODO(), clientId)
+		if err != nil {
+			flash.Error("OAuth client credential not found!")
+		} else {
+			c.Data["ClientId"] = oauthClient.GetID()
+			c.Data["ClientSecret"] = oauthClient.GetSecret()
+		}
+	}
+
+	flash.Store(&c.Controller)
+	c.Data["XSRFForm"] = template.HTML(c.XSRFFormHTML())
+	c.Layout = "layouts/application.html"
```

Comment:
As this out default layout, could we move it to the base controller so that we don't have to set it every time when we create a new controller?

Reply:
Moved to base controller